### PR TITLE
python312Packages.walrus: fix python 3.12 build

### PR DESCRIPTION
**Pull request by javaes**
## Description of changes

`distutils` import was deprecated in python 3.12. I am applying a patch derived from unreleased upstream which does the distutils related import in a safe way providing a 3.12 fallback. This should be removed in an upcoming release.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by RaghavSood** [comment]
Result of `nixpkgs-review pr 312546` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.walrus</li>
    <li>python311Packages.walrus.dist</li>
    <li>python312Packages.walrus</li>
    <li>python312Packages.walrus.dist</li>
  </ul>
</details>

**Comment by RaghavSood** [comment]
@GrahamcOfBorg build python311Packages.walrus python312Packages.walrus

**Comment by RaghavSood** [comment]
Result of `nixpkgs-review pr 312546` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.walrus</li>
    <li>python311Packages.walrus.dist</li>
    <li>python312Packages.walrus</li>
    <li>python312Packages.walrus.dist</li>
  </ul>
</details>

**Comment by RaghavSood** [comment]
Build failure on aarch64-darwin

[walrus.log](https://github.com/NixOS/nixpkgs/files/15360622/walrus.log)

**Comment by javaes** [comment]
Hi @RaghavSood , could it be possible that the Redis which is being used in the `aarch64-darwin` build is not properly sandboxed?

The tests cases should spawn their own Redis instance, but there seems to be one running already:
```
39751:M 18 May 2024 01:27:51.972 # Warning: Could not create server TCP listening socket *:6379: bind: Address already in use
```

**Comment by RaghavSood** [comment]
It is possible - I don't have anything else listening on that port, but maybe the sandboxing is off

Ofborg seems to have the same issue: https://logs.ofborg.org/?key=nixos/nixpkgs.312546&attempt_id=c67bb153-08c5-4637-a829-2758b3a6309c

**Comment by javaes** [comment]
I have now pushed a change which spawns the redis server on a custom port. Not sure why the darwin builds conflict with an already existing redis instance, let's see how it behaves now.

**Comment by RaghavSood** [comment]
@GrahamcOfBorg build python311Packages.walrus python312Packages.walrus

File: pkgs/development/python-modules/walrus/default.nix [hunk 1]
```diff
@@ -30,8 +38,9 @@ buildPythonPackage rec {
   ];
 
   preCheck = ''
-    ${pkgs.redis}/bin/redis-server &
+    ${pkgs.redis}/bin/redis-server --port 7777 &
     REDIS_PID=$!
+    export WALRUS_REDIS_PORT=7777
```
Comment (drupol) [inline pkgs/development/python-modules/walrus/default.nix:43]:
```suggestion
    export WALRUS_REDIS_PORT=7777
    ${pkgs.redis}/bin/redis-server --port $WALRUS_REDIS_PORT &
    REDIS_PID=$!
```

File: pkgs/development/python-modules/walrus/default.nix [hunk 2]
```diff
@@ -1,4 +1,5 @@
 { lib
+, stdenv
```
Comment (drupol) [inline pkgs/development/python-modules/walrus/default.nix:2]: Why this change?

File: pkgs/development/python-modules/walrus/default.nix [hunk 3]
```diff
@@ -21,6 +22,13 @@ buildPythonPackage rec {
     hash = "sha256-jinYMGSBAY8HTg92qU/iU5vGIrrDr5SeQG0XjsBVfcc=";
   };
 
+  patches = [
+    # distutils has been deprecated, this wraps its import inside a try-catch
+    # and fallsback to a fallback import.
+    # Should not be necessary in future versions.
+    ./fix-distutils-import.patch
+  ];
```
Comment (natsukium) [inline pkgs/development/python-modules/walrus/default.nix:30]: Why don't you use this commit instead?
https://github.com/coleifer/walrus/commit/79e20c89aa4015017ef8a3e0b5c27ca2731dc9b2

**Review by natsukium — COMMENTED** [review]
We can use `__darwinAllowLocalNetworking = true` here.

File: pkgs/development/python-modules/walrus/default.nix (hunk 2 above)
Comment (javaes) [reply pkgs/development/python-modules/walrus/default.nix]: It was a leftover from an in between state.

**Comment by drupol** [comment]
I've set this PR to draft because you removed a file in your last force push. Remove the draft flag when it's fixed and I'll give another review.

File: pkgs/development/python-modules/walrus/default.nix [hunk 4]
```diff
@@ -30,7 +41,8 @@ buildPythonPackage rec {
   ];
 
   preCheck = ''
-    ${pkgs.redis}/bin/redis-server &
+    export WALRUS_REDIS_PORT=7777
```
Comment (drupol) [inline pkgs/development/python-modules/walrus/default.nix:44]: Why is this environment variable needed?

Comment (javaes) [reply pkgs/development/python-modules/walrus/default.nix:44]: The redis-server failed to bind on the default port on darwin, this configures the walrus test case to run against the custom port.

Comment (drupol) [reply pkgs/development/python-modules/walrus/default.nix:44]: Can't we just get rid of it and use `7777` just under that line?

Comment (javaes) [reply pkgs/development/python-modules/walrus/default.nix:44]: It's there to configure the upstream test cases: https://github.com/coleifer/walrus/blob/0.9.3/walrus/tests/base.py#L9

Comment (drupol) [reply pkgs/development/python-modules/walrus/default.nix:44]: That's the thing. Can you please add a comment for that? Someone might remove it by mistake.

**Review by drupol — APPROVED** [review]
Thank you !

File: pkgs/development/python-modules/walrus/default.nix (hunk 4 above)
Comment (drupol) [reply pkgs/development/python-modules/walrus/default.nix:46]: ?

Comment (natsukium) [reply pkgs/development/python-modules/walrus/default.nix:46]:
```diff
- export WALRUS_REDIS_PORT=7777
- ${pkgs.redis}/bin/redis-server --port $WALRUS_REDIS_PORT &
+ ${pkgs.redis}/bin/redis-server &
```

**Review by natsukium — CHANGES_REQUESTED** [review]
The commit message and the PR title should start with `python312Packages.walrus` since this is a python library.
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-guidelines-contributing-guidelines

File: pkgs/development/python-modules/walrus/default.nix (hunk 4 above)
Comment (javaes) [reply pkgs/development/python-modules/walrus/default.nix]: reverted

**Review by natsukium — APPROVED** [review]
LGTM, thanks for your work!
I'm waiting for the CI.
